File: program/program/convert_string_to_sentencecase.pl
#!/usr/bin/perl

use strict;
use warnings;

sub sentence_case_with_replacement {
    my ($input, @fixed_words) = @_;
    my @sentences = split(/(\.|!|\?)\s+/, $input); # Split the input into sentences
    
    foreach my $sentence (@sentences) {
        $sentence =~ s/^\s+//; # Remove leading whitespace
        $sentence =~ s/\s+$//; # Remove trailing whitespace
        $sentence = ucfirst(lc($sentence)); # Convert to sentence case
        
        # Perform word replacements
        foreach my $word (@fixed_words) {
            $sentence =~ s/\b\Q$word\E\b/$word/gi; # Replace case-insensitively
        }
    }
    
    my $output = join(' ', @sentences); # Recombine the sentences with spaces
    
    return $output;
}

# Example 
my $input_string1 = "john and jane went to new york to visit the museum of modern art (moma).";
my @fixed_words1 = ("John", "Jane","New York", "Museum of Modern Art", "MoMA");

my $output_string1 = sentence_case_with_replacement($input_string1, @fixed_words1);

print $output_string1 . "\n";

my $input_string2 = "the united nations (un) is an international organization.";
my @fixed_words2 = ("The United Nations", "UN");

my $output_string2 = sentence_case_with_replacement($input_string2, @fixed_words2);

print $output_string2 . "\n";

my $input_string3 = "the cat is sleeping; however, the dog is awake.";
my @fixed_words3 = ();

my $output_string3 = sentence_case_with_replacement($input_string3, @fixed_words3);

print $output_string3 . "\n";

File: program/program/convert-string-to-sentencecase/convert_string_to_sentencecase.dart
String convertToSentenceCase(String input, List<String> fixedWords) {
  // Converting the input to lowercase
  input = input.toLowerCase();

  // Capitalizing the first character
  if (input.isNotEmpty) {
    input = input[0].toUpperCase() + input.substring(1);
  }
  else {
    return("Empty string");
  }

  // Replacing lower case fixed words with upper case
  for (String fixedWord in fixedWords) {
    String fixedWordLowerCase = fixedWord.toLowerCase();
    if (input.contains(fixedWordLowerCase)) {
      input = input.replaceAll(fixedWordLowerCase, fixedWord);
    }
  }

  // Handling special cases for colon, semicolon, and full stop
  for (int i = 0; i < input.length - 1; i++) {
    if (input[i] == ':' || input[i] == ';' || input[i] == '.') {
      int j = i + 1;
      // Skip any whitespace
      while (j < input.length && input[j] == ' ') {
        j++;
      }
      if (j < input.length) {
        input = input.substring(0, j) + input[j].toUpperCase() + input.substring(j + 1);
      }
    }
  }

  return input;
}

void main() {
  String text1 = "john and jane went to new york to visit the museum of modern art (moma)";
  List<String> fixedWords1 = ["John", "Jane", "New York", "Museum of Modern Art", "MoMA"];
  print(convertToSentenceCase(text1, fixedWords1));

  String text2 = "the united nations (un) is an international organization";
  List<String> fixedWords2 = ["United Nations", "UN"];
  print(convertToSentenceCase(text2, fixedWords2));

  String text3 = "the cat is sleeping; however, the dog is awake";
  List<String> fixedWords3 = [];
  print(convertToSentenceCase(text3, fixedWords3));
}
